src/Symfony/Component/Security/Core/Authentication/Token/AccessToken.php

```diff
@@ -0,0 +1,60 @@
+<?php
+
+/*
+ * This file is part of the Symfony package.
+ *
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Component\Security\Core\Authentication\Token;
+
+use Symfony\Component\Security\Core\User\UserInterface;
+
+/**
+ * AccessToken implements an access token.
+ * This type of token is commonly used in API contexts.
+ *
+ * @author Florent Morselli <florent.morselli@spomky-labs.com>
+ */
+class AccessToken extends AbstractToken
+{
+    private string $firewallName;
+
+    public function __construct(UserInterface $user, string $firewallName, array $roles = [])
+    {
+        parent::__construct($roles);
+
+        if ('' === $firewallName) {
+            throw new \InvalidArgumentException('$firewallName must not be empty.');
+        }
+
+        $this->setUser($user);
+        $this->firewallName = $firewallName;
+    }
+
+    public function getFirewallName(): string
+    {
+        return $this->firewallName;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function __serialize(): array
+    {
+        return [null, $this->firewallName, parent::__serialize()];
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function __unserialize(array $data): void
+    {
+        [, $this->firewallName, $parentData] = $data;
+        $parentData = \is_array($parentData) ? $parentData : unserialize($parentData);
```

Comment:
Why is this check necessary? This should always be an array AFAIU.

Reply:
Well... I am not sure from where it comes. I replicated this part from other Token classes such as RememberMeToken or SwitchUserToken.
I wanted to be consistent with the other ones.

Reply:
IIRC this check is about handling session tokens serialized in Symfony 3.x in order to not break hard when upgrading from e.g. 3.4 to 4.4 or higher, given `$parentData` used to be a string before 4.x.